[DTensor] Add single-dim registration infra

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170827
* __->__ #170359
* #167677
* #170615


This PR adds the register_single_dim_strategy util,  and hooks it up to sharding_propagator.  It also tests the registration.

Notes:
* I didn't yet decide how multiple registrations should be handled.  I was planning to make it an error if you register twice for the same op for either single_dim or regular strategies.
* I took the cleanest path of integration for now in sharding_prop, reusing as much code as possible with the existing 'op_strategy' case.  I may have to change this later when integrating find_min_cost

wconstab committed:
[DTensor] Add single-dim registration infra

[ghstack-poisoned]

pianpwk committed:
Update

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add single-dim registration infra"




[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add single-dim registration infra"




[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add single-dim registration infra"




[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add single-dim registration infra"



This PR adds the register_single_dim_strategy util,  and hooks it up to sharding_propagator.  It also tests the registration.

Notes:
* I didn't yet decide how multiple registrations should be handled.  I was planning to make it an error if you register twice for the same op for either single_dim or regular strategies.
* I took the cleanest path of integration for now in sharding_prop, reusing as much code as possible with the existing 'op_strategy' case.  I may have to change this later when integrating find_min_cost 



[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] Add single-dim registration infra"



This PR adds the register_single_dim_strategy util,  and hooks it up to sharding_propagator.  It also tests the registration.

Notes:
* I didn't yet decide how multiple registrations should be handled.  I was planning to make it an error if you register twice for the same op for either single_dim or regular strategies.
* I took the cleanest path of integration for now in sharding_prop, reusing as much code as possible with the existing 'op_strategy' case.  I may have to change this later when integrating find_min_cost 



[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

jeanschmidt commented:
@pytorchbot revert -m "Required to revert https://github.com/pytorch/pytorch/pull/167677 that is required to revert https://github.com/pytorch/pytorch/pull/170615 that is required to revert https://github.com/pytorch/pytorch/pull/170030" -c ghfirst

jeanschmidt commented:
this is the most disruptive revert I ever did, not great :(

wconstab committed:
Update on "[DTensor] Add single-dim registration infra"



This PR adds the register_single_dim_strategy util,  and hooks it up to sharding_propagator.  It also tests the registration.

Notes:
* I didn't yet decide how multiple registrations should be handled.  I was planning to make it an error if you register twice for the same op for either single_dim or regular strategies.
* I took the cleanest path of integration for now in sharding_prop, reusing as much code as possible with the existing 'op_strategy' case.  I may have to change this later when integrating find_min_cost 



[ghstack-poisoned]

wconstab commented:
@pytorchbot merge -i
